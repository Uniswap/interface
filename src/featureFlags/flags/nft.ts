import { BaseVariant, FeatureFlag, useBaseFlag } from '../index'

export function useNftFlag(): BaseVariant {
  return useBaseFlag(FeatureFlag.nft)
}

export { BaseVariant as NftVariant }
