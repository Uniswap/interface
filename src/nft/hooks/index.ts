export * from './useBag'
export * from './useCollectionFilters'
export * from './useFiltersExpanded'
export * from './useIsCollectionLoading'
export * from './useIsMobile'
export * from './useIsTablet'
export * from './useMarketplaceSelect'
export * from './useNFTList'
export * from './useNFTSelect'
export * from './useProfilePageState'
export * from './useSearchHistory'
export * from './useSelectAsset'
export * from './useSellAsset'
export * from './useSendTransaction'
export * from './useSweep'
export * from './useTransactionResponse'
export * from './useWalletBalance'
export * from './useWalletCollections'
