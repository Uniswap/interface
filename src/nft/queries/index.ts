export * from './genie'
export * from './looksRare'
export * from './openSea'
export * from './x2y2'
