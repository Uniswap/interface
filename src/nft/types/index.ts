export * from './checkout'
export * from './collection'
export * from './common'
export * from './discover'
export * from './navbar'
export * from './sell'
