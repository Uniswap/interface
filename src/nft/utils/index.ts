export * from './asset'
export * from './buildActivityAsset'
export * from './buildSellObject'
export * from './calcPoolPrice'
export * from './currency'
export * from './fetchPrice'
export * from './isAudio'
export * from './isVideo'
export * from './listNfts'
export * from './putCommas'
export * from './rarity'
export * from './roundAndPluralize'
export * from './timeSince'
export * from './transactionResponse'
export * from './updatedAssets'
