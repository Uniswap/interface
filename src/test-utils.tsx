import { render, RenderOptions } from '@testing-library/react'
import React, { FC, ReactElement, ReactNode } from 'react'
import { Provider } from 'react-redux'
import store from 'state'
import ThemeProvider from 'theme'

const WithProviders: FC = ({ children }: { children?: ReactNode }) => {
  return (
    <Provider store={store}>
      <ThemeProvider>{children}</ThemeProvider>
    </Provider>
  )
}

const customRender = (ui: ReactElement, options?: Omit<RenderOptions, 'wrapper'>) =>
  render(ui, { wrapper: WithProviders, ...options })

export * from '@testing-library/react'
export { customRender as render }
