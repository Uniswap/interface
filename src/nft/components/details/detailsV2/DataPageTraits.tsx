import { Trans } from '@lingui/macro'
import Column from 'components/Column'
import { ScrollBarStyles } from 'components/Common'
import Row from 'components/Row'
import { useSubscribeScrollState } from 'nft/hooks'
import { GenieAsset } from 'nft/types'
import { useMemo } from 'react'
import styled from 'styled-components/macro'
import { BREAKPOINTS, ThemedText } from 'theme'

import { Scrim } from './shared'
import { Tab, TabbedComponent } from './TabbedComponent'
import { TraitRow } from './TraitRow'

const TraitsHeaderContainer = styled(Row)`
  padding: 0px 12px;
`

const TraitsHeader = styled(ThemedText.SubHeaderSmall)<{
  $flex?: number
  $justifyContent?: string
  hideOnSmall?: boolean
}>`
  display: flex;
  line-height: 20px;
  color: ${({ theme }) => theme.textSecondary};
  flex: ${({ $flex }) => $flex ?? 1};
  justify-content: ${({ $justifyContent }) => $justifyContent};

  @media screen and (max-width: ${BREAKPOINTS.sm}px) {
    display: ${({ hideOnSmall }) => (hideOnSmall ? 'none' : 'flex')};
  }
`

const TraitRowContainer = styled.div`
  position: relative;
`

const TraitRowScrollableContainer = styled.div`
  overflow-y: auto;
  overflow-x: hidden;
  max-height: 412px;
  width: calc(100% + 6px);

  ${ScrollBarStyles}
`

const TraitsContent = ({ asset }: { asset: GenieAsset }) => {
  const { userCanScroll, scrollRef, scrollProgress, scrollHandler } = useSubscribeScrollState()

  // This is needed to prevent rerenders when handling scrolls
  const traitRows = useMemo(() => {
    return asset.traits?.map((trait) => (
      <TraitRow collectionAddress={asset.address} trait={trait} key={trait.trait_type + ':' + trait.trait_value} />
    ))
  }, [asset.address, asset.traits])

  return (
    <Column>
      <TraitsHeaderContainer>
        <TraitsHeader $flex={3}>
          <Trans>Trait</Trans>
        </TraitsHeader>
        <TraitsHeader $flex={2}>
          <Trans>Floor price</Trans>
        </TraitsHeader>
        <TraitsHeader hideOnSmall={true}>
          <Trans>Quantity</Trans>
        </TraitsHeader>
        <TraitsHeader $flex={1.5} $justifyContent="flex-end">
          <Trans>Rarity</Trans>
        </TraitsHeader>
      </TraitsHeaderContainer>
      <TraitRowContainer>
        {scrollProgress > 0 && <Scrim />}
        <TraitRowScrollableContainer ref={scrollRef} onScroll={scrollHandler}>
          {traitRows}
        </TraitRowScrollableContainer>
        {userCanScroll && scrollProgress !== 100 && <Scrim isBottom={true} />}
      </TraitRowContainer>
    </Column>
  )
}

enum TraitTabsKeys {
  Traits = 'traits',
}

export const DataPageTraits = ({ asset }: { asset: GenieAsset }) => {
  const TraitTabs: Map<string, Tab> = useMemo(
    () =>
      new Map([
        [
          TraitTabsKeys.Traits,
          {
            title: <Trans>Traits</Trans>,
            key: TraitTabsKeys.Traits,
            content: <TraitsContent asset={asset} />,
            count: asset.traits?.length,
          },
        ],
      ]),
    [asset]
  )
  return <TabbedComponent tabs={TraitTabs} />
}
