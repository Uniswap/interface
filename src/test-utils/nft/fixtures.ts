import {
  MediaType,
  NftActivityType,
  NftMarketplace,
  NftStandard,
  OrderStatus,
  OrderType,
} from 'graphql/data/__generated__/types-and-hooks'
import ms from 'ms.macro'
import { ActivityEvent, CollectionInfoForAsset, GenieAsset, Markets, Offer, SellOrder, WalletAsset } from 'nft/types'

export const TEST_NFT_ASSET: GenieAsset = {
  id: 'TmZ0QXNzZXQ6MHhlZDVhZjM4ODY1MzU2N2FmMmYzODhlNjIyNGRjN2M0YjMyNDFjNTQ0XzMzMTg=',
  address: '0xed5af388653567af2f388e6224dc7c4b3241c544',
  notForSale: false,
  collectionName: 'Azuki',
  imageUrl:
    'https://cdn.center.app/1/0xED5AF388653567Af2F388E6224dC7C4b3241C544/3318/50ed67ad647d0aa0cad0b830d136a677efc2fb72a44587bc35f2a5fb334a7fdf.png',
  mediaType: MediaType.Image,
  marketplace: Markets.Opensea,
  name: 'Azuki #3318',
  priceInfo: {
    ETHPrice: '15800000000000000000',
    baseAsset: 'ETH',
    baseDecimals: '18',
    basePrice: '15800000000000000000',
  },
  susFlag: false,
  tokenId: '3318',
  tokenType: NftStandard.Erc721,
  totalCount: 10000,
  collectionIsVerified: true,
  rarity: {
    primaryProvider: 'Rarity Sniper',
    providers: [
      {
        rank: 7079,
        provider: 'Rarity Sniper',
      },
    ],
  },
  creator: {},
}

export const TEST_VIDEO_NFT_ASSET: GenieAsset = {
  id: 'TmZ0QXNzZXQ6MHg0ZTFmNDE2MTNjOTA4NGZkYjllMzRlMTFmYWU5NDEyNDI3NDgwZTU2XzcyMDI=',
  address: '0x6c9369dc930fd794ad0af7511f483d936a2ef7f3',
  notForSale: false,
  collectionName: 'Terraforms by Mathcastles',
  imageUrl:
    'https://i.seadn.io/gae/tkDbNhjjBZV2PmYaJbJOOigywZCrlcyGRxeQFkZS1YZyihyG5GoWNWj3N9f1T7YVuaxOqdxhfJylC9ejtoCvdgBE932vd7jorVqA?w=500&auto=format',
  animationUrl: 'https://openseauserdata.com/files/5af92728200027caa4f3f5ae87a486a7.mp4',
  mediaType: MediaType.Video,
  marketplace: Markets.Opensea,
  name: 'Aku Chapter IV: Aku x Ady #884',
  priceInfo: {
    ETHPrice: '1295000000000000000',
    baseAsset: 'ETH',
    baseDecimals: '18',
    basePrice: '1295000000000000000',
  },
  susFlag: false,
  tokenId: '1455',
  tokenType: NftStandard.Erc721,
  collectionIsVerified: false,
  totalCount: 9910,
  rarity: {
    primaryProvider: 'Rarity Sniper',
    providers: [
      {
        rank: 7079,
        provider: 'Rarity Sniper',
      },
    ],
  },
  creator: {},
}

export const TEST_AUDIO_NFT_ASSET: GenieAsset = {
  id: 'TmZ0QXNzZXQ6MHg0ZTFmNDE2MTNjOTA4NGZkYjllMzRlMTFmYWU5NDEyNDI3NDgwZTU2XzcyMDI=',
  address: '0x37a03d4af1d7046d1126987b20117a0fdcbf6535',
  notForSale: false,
  collectionName: 'Snoop Dogg on Sound XYZ',
  imageUrl:
    'https://i.seadn.io/gae/Kze9SBqn_6O0qrHKxspo1gRkkDV2A5EmTeWtvdS-dNxBsvi_wPXUYjc6De0sUC-DYzL093102mUftenWxwWuTelqsdw-ngoBC3o2XFU?w=500&auto=format',
  animationUrl: 'https://openseauserdata.com/files/4a22253e44e10baa11484a2e43efefda.mp3',
  mediaType: MediaType.Audio,
  marketplace: Markets.Opensea,
  name: 'Death Row Session: Vol. 2 (420 Edition) #320',
  priceInfo: {
    ETHPrice: '1295000000000000000',
    baseAsset: 'ETH',
    baseDecimals: '18',
    basePrice: '1295000000000000000',
  },
  susFlag: false,
  tokenId: '680564733841876926926749214863536423232',
  tokenType: NftStandard.Erc721,
  collectionIsVerified: false,
  totalCount: 9910,
  rarity: {
    primaryProvider: 'Rarity Sniper',
    providers: [
      {
        rank: 7079,
        provider: 'Rarity Sniper',
      },
    ],
  },
  creator: {},
}

export const TEST_EMBEDDED_NFT_ASSET: GenieAsset = {
  id: 'TmZ0QXNzZXQ6MHg0ZTFmNDE2MTNjOTA4NGZkYjllMzRlMTFmYWU5NDEyNDI3NDgwZTU2XzcyMDI=',
  address: '0x4e1f41613c9084fdb9e34e11fae9412427480e56',
  notForSale: false,
  collectionName: 'Terraforms by Mathcastles',
  imageUrl:
    'https://cdn.center.app/v2/1/06ff92279474add6ce06176e2a65447396edf786d169d8ccc03fddfa45ce004f/bb01f8a2f093ea4619498dae58fc19e5ba3fa38a84cabf92948994609489d566.png',
  animationUrl: 'https://tokens.mathcastles.xyz/terraforms/token-html/7202',
  mediaType: MediaType.Raw,
  marketplace: Markets.Opensea,
  name: 'Level 13 at {28, 3}',
  priceInfo: {
    ETHPrice: '1295000000000000000',
    baseAsset: 'ETH',
    baseDecimals: '18',
    basePrice: '1295000000000000000',
  },
  susFlag: false,
  tokenId: '7202',
  tokenType: NftStandard.Erc721,
  collectionIsVerified: false,
  totalCount: 9910,
  rarity: {
    primaryProvider: 'Rarity Sniper',
    providers: [
      {
        rank: 7079,
        provider: 'Rarity Sniper',
      },
    ],
  },
  creator: { address: '0xe72eb31b59f85b19499a0f3b3260011894fa0d65' },
}

export const TEST_NFT_WALLET_ASSET: WalletAsset = {
  id: 'TmZ0QXNzZXQ6RVRIRVJFVU1fMHgyOTY1MkMyZTlEMzY1NjQzNEJjODEzM2M2OTI1OEM4ZDA1MjkwZjQxXzIzNTk=',
  imageUrl: 'https://c.neevacdn.net/image/upload/xyz/T96PksTnWGNh79CrzLn-zpYfqRWtD5wME0MBPL_Md6Q.png',
  smallImageUrl:
    'https://c.neevacdn.net/image/upload/c_limit,pg_1,h_1200,w_1200/f_webp/xyz/T96PksTnWGNh79CrzLn-zpYfqRWtD5wME0MBPL_Md6Q.webp',
  notForSale: true,
  priceInfo: {
    ETHPrice: '0',
    baseAsset: 'ETH',
    baseDecimals: '18',
    basePrice: '0',
  },
  name: 'Froggy Friend #2359',
  tokenId: '2359',
  asset_contract: {
    address: '0x29652c2e9d3656434bc8133c69258c8d05290f41',
    tokenType: NftStandard.Erc721,
    name: 'Froggy Friends Official',
    description: '4444 of the friendliest frogs in the metaverse.',
    image_url: 'https://i.seadn.io/gcs/files/84483786d97b4d471cb48d224c4c5c91.png?w=500&auto=format',
  },
  collection: {
    address: '0x29652c2e9d3656434bc8133c69258c8d05290f41',
    name: 'Froggy Friends Official',
    isVerified: true,
    imageUrl: 'https://i.seadn.io/gcs/files/84483786d97b4d471cb48d224c4c5c91.png?w=500&auto=format',
    twitterUrl: '@FroggyFriendNFT',
  },
  collectionIsVerified: true,
  lastPrice: 0,
  floorPrice: 0.0775,
  basisPoints: 0,
  date_acquired: '1682024661',
  sellOrders: [],
}

export const TEST_NFT_ACTIVITY_EVENT: ActivityEvent = {
  collectionAddress: '0xed5af388653567af2f388e6224dc7c4b3241c544',
  tokenId: '5674',
  tokenMetadata: {
    name: 'Azuki #5674',
    imageUrl:
      'https://cdn.center.app/1/0xED5AF388653567Af2F388E6224dC7C4b3241C544/5674/b2e5cb241d4a28bb3688ff6ae12f2d60c9850721f35f5104b5c42b31511e8a42.png',
    smallImageUrl: 'https://i.seadn.io/gcs/files/e2dabe8f353ed6354f5a1927e3d8bd64.png?w=500&auto=format',
    metadataUrl: 'ipfs://QmZcH4YvBVVRJtdn4RdbaqgspFU8gH6P9vomDpBVpAL3u4/5674',
    rarity: {
      source: 'RARITY_SNIPER',
      rank: 9412,
      score: 2778,
    },
    suspiciousFlag: false,
    standard: NftStandard.Erc721,
  },
  eventType: NftActivityType.Listing,
  marketplace: 'OPENSEA',
  fromAddress: '0xbf9fda32692b25c6083cbe48399ef019b62f0712',
  toAddress: undefined,
  transactionHash: undefined,
  price: '15.2',
  orderStatus: OrderStatus.Valid,
  quantity: 1,
  url: 'https://opensea.io/assets/0xed5af388653567af2f388e6224dc7c4b3241c544/5674',
  eventTimestamp: 1682444662,
}

export const TEST_NFT_COLLECTION_INFO_FOR_ASSET: CollectionInfoForAsset = {
  collectionDescription:
    'Take the red bean to join the garden. View the collection atÂ [azuki.com/gallery](https://azuki.com/gallery).\r\n\r\nAzuki starts with a collection of 10,000 avatars that give you membership access to The Garden: a corner of the internet where artists, builders, and web3 enthusiasts meet to create a decentralized future. Azuki holders receive access to exclusive drops, experiences, and more. Visit [azuki.com](https://azuki.com) for more details.\r\n\r\nWe rise together. We build together. We grow together.',
  collectionImageUrl:
    'https://i.seadn.io/gae/H8jOCJuQokNqGBpkBN5wk1oZwO7LM8bNnrHCaekV2nKjnCqw6UB5oaH8XyNeBDj6bA_n1mjejzhFQUP3O1NfjFLHr3FOaeHcTOOT?w=500&auto=format',
  collectionName: 'Azuki',
  discordUrl: 'https://discord.gg/azuki',
  externalUrl: 'http://www.azuki.com',
  isVerified: true,
  totalSupply: 10000,
}

const FIVE_MONTHS_MILLISECONDS = ms`165 days`

export const TEST_SELL_ORDER: SellOrder = {
  address: '0x29d7ebca656665c1a52a92f830e413e394db6b4f',
  createdAt: 1683561510000,
  endAt: Date.now() + FIVE_MONTHS_MILLISECONDS,
  id: 'TmZ0T3JkZXI6MHgyOWQ3ZWJjYTY1NjY2NWMxYTUyYTkyZjgzMGU0MTNlMzk0ZGI2YjRmXzY4MTVfMHg3OWVhNDQ5YzMzNzVlZDFhOWQ3ZDk5ZjgwNjgyMDllYTc0OGM2ZDQyXzQ5NzAwMDAwMDAwMDAwMDAwMDAwMF9vcGVuc2VhX01vbiBNYXkgMDggMjAyMyAxNTo1ODozMCBHTVQrMDAwMCAoQ29vcmRpbmF0ZWQgVW5pdmVyc2FsIFRpbWUp',
  maker: '0x79ea449c3375ed1a9d7d99f8068209ea748c6d42',
  marketplace: NftMarketplace.Opensea,
  marketplaceUrl: 'https://opensea.io/assets/0x29d7ebca656665c1a52a92f830e413e394db6b4f/6815',
  price: {
    currency: 'ETH',
    value: 99999999,
  },
  quantity: 1,
  startAt: 1683561507000,
  status: OrderStatus.Valid,
  type: OrderType.Listing,
  protocolParameters: {},
}

export const TEST_OFFER: Offer = {
  createdAt: 1683561510000,
  endAt: Date.now() + FIVE_MONTHS_MILLISECONDS,
  id: 'TmZ0T3JkZXI6MHgyOWQ3ZWJjYTY1NjY2NWMxYTUyYTkyZjgzMGU0MTNlMzk0ZGI2YjRmXzY4MTVfMHg3OWVhNDQ5YzMzNzVlZDFhOWQ3ZDk5ZjgwNjgyMDllYTc0OGM2ZDQyXzQ5NzAwMDAwMDAwMDAwMDAwMDAwMF9vcGVuc2VhX01vbiBNYXkgMDggMjAyMyAxNTo1ODozMCBHTVQrMDAwMCAoQ29vcmRpbmF0ZWQgVW5pdmVyc2FsIFRpbWUp',
  maker: '0x79ea449c3375ed1a9d7d99f8068209ea748c6d42',
  marketplace: NftMarketplace.Opensea,
  marketplaceUrl: 'https://opensea.io/assets/0x29d7ebca656665c1a52a92f830e413e394db6b4f/6815',
  price: {
    currency: 'ETH',
    value: 123.456,
  },
  quantity: 1,
}
