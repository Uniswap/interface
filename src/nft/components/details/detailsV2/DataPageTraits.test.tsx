import { TEST_NFT_ASSET } from 'test-utils/nft/fixtures'
import { render } from 'test-utils/render'

import { DataPageTraits } from './DataPageTraits'

it('data page trait component does not load with asset with no traits', () => {
  const { asFragment } = render(<DataPageTraits asset={TEST_NFT_ASSET} />)
  expect(asFragment()).toMatchSnapshot()
})

// TODO(NFT-1189): add test for trait component with asset with traits when rarity is not randomly generated
// while rarities are randomly generated, snapshots will never match
