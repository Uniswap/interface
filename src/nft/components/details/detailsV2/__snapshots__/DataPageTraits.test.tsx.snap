// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`data page trait component does not load with asset with no traits 1`] = `
<DocumentFragment>
  .c1 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
}

.c2 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 0;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
}

.c6 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 0;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  gap: 8px;
}

.c4 {
  color: #0D111C;
}

.c9 {
  color: #7780A0;
}

.c7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
}

.c0 {
  background: #FFFFFF;
  border: 1px solid #D2D9EE;
  border-radius: 16px;
  padding: 16px 20px;
  width: 100%;
  -webkit-align-self: flex-start;
  -ms-flex-item-align: start;
  align-self: flex-start;
}

.c3 {
  gap: 32px;
  width: 100;
  margin-bottom: 12px;
  padding-bottom: 12px;
  border-bottom: 1px solid #D2D9EE;
}

.c5 {
  color: #0D111C;
  line-height: 24px;
  cursor: default;
}

.c8 {
  padding: 0px 12px;
}

.c10 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  line-height: 20px;
  color: #7780A0;
  -webkit-flex: 3;
  -ms-flex: 3;
  flex: 3;
}

.c11 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  line-height: 20px;
  color: #7780A0;
  -webkit-flex: 2;
  -ms-flex: 2;
  flex: 2;
}

.c12 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  line-height: 20px;
  color: #7780A0;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c13 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  line-height: 20px;
  color: #7780A0;
  -webkit-flex: 1.5;
  -ms-flex: 1.5;
  flex: 1.5;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
}

.c14 {
  position: relative;
}

.c15 {
  overflow-y: auto;
  overflow-x: hidden;
  max-height: 412px;
  width: calc(100% + 6px);
  -webkit-scrollbar-width: thin;
  -moz-scrollbar-width: thin;
  -ms-scrollbar-width: thin;
  scrollbar-width: thin;
  -webkit-scrollbar-color: #D2D9EE transparent;
  -moz-scrollbar-color: #D2D9EE transparent;
  -ms-scrollbar-color: #D2D9EE transparent;
  scrollbar-color: #D2D9EE transparent;
  height: 100%;
}

.c15::-webkit-scrollbar {
  background: transparent;
  width: 4px;
  overflow-y: scroll;
}

.c15::-webkit-scrollbar-thumb {
  background: #D2D9EE;
  border-radius: 8px;
}

@media screen and (max-width:640px) {
  .c10 {
    display: -webkit-box;
    display: -webkit-flex;
    display: -ms-flexbox;
    display: flex;
  }
}

@media screen and (max-width:640px) {
  .c11 {
    display: -webkit-box;
    display: -webkit-flex;
    display: -ms-flexbox;
    display: flex;
  }
}

@media screen and (max-width:640px) {
  .c12 {
    display: none;
  }
}

@media screen and (max-width:640px) {
  .c13 {
    display: -webkit-box;
    display: -webkit-flex;
    display: -ms-flexbox;
    display: flex;
  }
}

<div
    class="c0"
  >
    <div
      class="c1 c2 c3"
    >
      <div
        class="c4 c5 css-rjqmed"
      >
        <div
          class="c1 c6"
        >
          Traits
        </div>
      </div>
    </div>
    <div
      class="c7"
    >
      <div
        class="c1 c2 c8"
      >
        <div
          class="c9 c10 css-1aekuku"
        >
          Trait
        </div>
        <div
          class="c9 c11 css-1aekuku"
        >
          Floor price
        </div>
        <div
          class="c9 c12 css-1aekuku"
        >
          Quantity
        </div>
        <div
          class="c9 c13 css-1aekuku"
        >
          Rarity
        </div>
      </div>
      <div
        class="c14"
      >
        <div
          class="c15"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
